src/comgr.cpp

```diff
@@ -0,0 +1,490 @@
+    }
+    Data GetData(const amd_comgr_data_kind_t kind, const std::size_t index) const
+    {
+        amd_comgr_data_t d;
+        ECI_THROW(amd_comgr_action_data_get_data(handle, kind, index, &d), kind);
+        return {d};
+    }
+};
+
+class ActionInfo : ComgrOwner
+{
+    amd_comgr_action_info_t handle = {0};
+
+    public:
+    ActionInfo() { EC_THROW(amd_comgr_create_action_info(&handle)); }
+    ~ActionInfo() { EC(amd_comgr_destroy_action_info(handle)); }
+    auto operator()() const { return handle; }
+    void SetLanguage(const amd_comgr_language_t language) const
+    {
+        ECI_THROW(amd_comgr_action_info_set_language(handle, language), language);
+    }
+    void SetIsaName(const std::string& isa) const
+    {
+        ECI_THROW_MSG(amd_comgr_action_info_set_isa_name(handle, isa.c_str()), isa, isa);
+    }
+    void SetLogging(const bool state) const
+    {
+        ECI_THROW(amd_comgr_action_info_set_logging(handle, state), state);
+    }
+    void SetOptionList(const std::vector<std::string>& options) const
+    {
+        std::vector<const char*> vp;
+        vp.reserve(options.size());
+        for(auto& opt : options) // cppcheck-suppress useStlAlgorithm
+            vp.push_back(opt.c_str());
+        LogOptions(vp.data(), vp.size());
+        ECI_THROW(amd_comgr_action_info_set_option_list(handle, vp.data(), vp.size()), vp.size());
+    }
+    void Do(const amd_comgr_action_kind_t kind, const Dataset& in, const Dataset& out) const
+    {
+        ECI_THROW_MSG(amd_comgr_do_action(kind, handle, in(), out()), kind, GetLog(out, true));
+        const auto log = GetLog(out);
+        if(!log.empty())
+            MIOPEN_LOG_I(to_string(kind) << ": " << log);
+    }
+};
+
+/// If called from the context of comgr error handling:
+/// - We do not allow the comgr-induced exceptions to escape.
+/// - If obtaining log fails, write some diagnostics into output.
+static std::string GetLog(const Dataset& dataset, const bool comgr_error_handling)
+{
+    std::string text;
+    try
+    {
+        /// Assumption: the log is the the first in the dataset.
+        /// This is not specified in comgr API.
+        /// Let's follow the KISS principle for now; it works.
+        ///
+        /// \todo Clarify API and update implementation.
+        const auto count = dataset.GetDataCount(AMD_COMGR_DATA_KIND_LOG);
+        if(count < 1)
+            return {comgr_error_handling ? "comgr warning: error log not found" : ""};
+
+        const auto data = dataset.GetData(AMD_COMGR_DATA_KIND_LOG, 0);
+        text            = data.GetString();
+        if(text.empty())
+            return {comgr_error_handling ? "comgr info: error log empty" : ""};
+    }
+    catch(ComgrError&)
+    {
+        if(comgr_error_handling)
+            return {"comgr error: failed to get error log"};
+        // deepcode ignore EmptyThrowOutsideCatch: false positive
+        throw;
+    }
+    return text;
+}
+
+void BuildOcl(const std::string& name,
```

Comment:
This function should return `std::vector<char>` and not use output parameters.

Reply:
If `GetBytes` does not return a temporary object, then returning `std::vector<char>` here would involve NRVO, which [I would like to avoid](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/177#discussion_r421128972).